docs-guides/source/stateful-models.md

```diff
@@ -227,6 +227,219 @@ potential runtime performance improvements.
+
+converted_model = ct.convert(
+     traced_model,
+     inputs=inputs,
+     outputs=outputs,
+     minimum_deployment_target=ct.target.iOS18,
+     compute_units=ct.ComputeUnit.CPU_AND_GPU,
+)
+```
+Notice that the `minimum_deployment_target=ct.target.iOS18` is not necessary if you only
+want to use the stateless model, as stateless models are supported before iOS18.
+Here we set it just for fair comparison with the stateful kvcache model later.
+
+We can time the prediction of the stateless model
+```python
+from time import perf_counter
+
+t_start: float = perf_counter()
+for token_id in range(0, num_iterations):
+     inputs = {"input_ids": np.array([list(range(token_id + 1))], dtype=np.int32)}
+     converted_model.predict(inputs)
+print(f"Time (ms) without kv-cache: {(perf_counter() - t_start) * 1.0e3}")
+```
+
+Now let's initialize and convert the stateful kv-cache model in a similar way
+```python
+past_kv_len = 0
+torch_model_kvcache = ToyModelWithKeyValueCache(
+    vocab_size, embed_size, batch_size, max_seq_len
+)
+torch_model_kvcache.load_state_dict(torch_model.state_dict(), strict=False)
+torch_model_kvcache.eval()
+causal_mask = torch.zeros((batch_size, seq_len, seq_len + past_kv_len), dtype=torch.float32)
+
+# Make sure the output matches the non-kv-cache version.
+torch_kvcache_output = torch_model_kvcache(input_ids, causal_mask).detach().numpy()
+np.testing.assert_allclose(torch_output, torch_kvcache_output)
+
+traced_model_kvcache = torch.jit.trace(torch_model_kvcache, [input_ids, causal_mask])
+query_length = ct.RangeDim(lower_bound=1, upper_bound=max_seq_len, default=1)
+end_step_dim = ct.RangeDim(lower_bound=1, upper_bound=max_seq_len, default=1)
+inputs = [
+    ct.TensorType(shape=(batch_size, query_length), dtype=np.int32, name="input_ids"),
+    ct.TensorType(
+        shape=(batch_size, query_length, end_step_dim), dtype=np.float16, name="causal_mask"
+    ),
+]
+outputs = [ct.TensorType(dtype=np.float16, name="output")]
+
+# In addition to `inputs` and `outputs`, we need `states` which uses the same name as the
+# registered buffers in `ToyModelWithKeyValueCache`.
+states = [
+    ct.StateType(
+        wrapped_type=ct.TensorType(
+            shape=torch_model_kvcache.kvcache_shape, dtype=np.float16
+        ),
+        name="k_cache",
+    ),
+    ct.StateType(
+        wrapped_type=ct.TensorType(
+            shape=torch_model_kvcache.kvcache_shape, dtype=np.float16
+        ),
+        name="v_cache",
+    ),
+]
+converted_model_kvcache = ct.convert(
+    traced_model_kvcache,
+    inputs=inputs,
+    outputs=outputs,
+    states=states,
+    minimum_deployment_target=ct.target.iOS18,
+    compute_units=ct.ComputeUnit.CPU_AND_GPU,
+)
+```
+
+We can also time the prediction of this stateful kv-cache model
+```python
+past_kv_len = 0
+kv_cache_state = converted_model_kvcache.make_state()
+t_start: float = perf_counter()
+for token_id in range(0, num_iterations):
```

Comment:
nit: `range(0, num_iterations)` is equivalent to `range(num_iterations)`

Reply:
Good catch! Changed.